infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/secure/SecureServerExposerFactory.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.server.secure;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.external.ExternalServerExposerStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.SecureServerExposer.DefaultSecureServerExposer;
+import org.eclipse.che.workspace.infrastructure.kubernetes.server.secure.jwtproxy.JwtProxySecureServerExposerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** @author Sergii Leshchenko */
+public class SecureServerExposerFactory<T extends KubernetesEnvironment> {
+  private static final Logger LOG = LoggerFactory.getLogger(SecureServerExposerFactory.class);
+
+  private final boolean agentsAuthEnabled;
+  private final String serverExposer;
+
+  private final ExternalServerExposerStrategy<T> exposerStrategy;
+  private final JwtProxySecureServerExposerFactory<T> jwtProxySecureServerExposerFactory;
+
+  @Inject
+  public SecureServerExposerFactory(
+      @Named("che.agents.auth_enabled") boolean agentsAuthEnabled,
+      @Named("che.agents.auth.secure_server_exposer") String serverExposer,
+      ExternalServerExposerStrategy<T> exposerStrategy,
+      JwtProxySecureServerExposerFactory<T> jwtProxySecureServerExposerFactory) {
+    this.agentsAuthEnabled = agentsAuthEnabled;
+    this.serverExposer = serverExposer;
+    this.exposerStrategy = exposerStrategy;
+    this.jwtProxySecureServerExposerFactory = jwtProxySecureServerExposerFactory;
+  }
+
+  /**
+   * Creates instance of {@link SecureServerExposer} that will expose secure servers for runtime
+   * with the specified runtime identity.
+   */
+  public SecureServerExposer<T> create(RuntimeIdentity identity) {
+    if (!agentsAuthEnabled) {
+      // return default secure server exposer because no need to protect servers with authentication
+      return new DefaultSecureServerExposer<>(exposerStrategy);
+    }
+
+    switch (serverExposer) {
+      case "jwtproxy":
```

Comment:
Makes sense, I'll rework it with Map bindings.

Reply:
Ouch. I missed that `JwtProxySecureServerExposer` is created with factory that uses assisted inject. So, SecureServerExposerFactory could inject something like `Map<String, SecureServerExposerImplFactory>` where `SecureServerExposerImplFactory` is a custom interface 
```
interface SecureServerExposerImplFactory {
  SecureServerExpose create(RuntimeIdentity identity);
}
```
And then there should be corresponding SecureServerExposerImplFactory implementaion of each secure server (now JwtProxy and Default).

Looks like a bit overkill for now.

Do you think I should rework it or it is OK to leave it as it is for now?

Reply:
Maybe we can use a single interface of a factory of a factory instead of the map) This would allow using another implementation instead of JWT one (it can be even customized version of JWT one, depending on the ISV needs).
But yes, if there is no elegant and easy to implement and support way or if it will delay merging of the PR we can totally return to this issue when there would be a need for that.
And it looks like we definitely need to leave the approach as it is implemented now.

Reply:
Reworked in the suggested way.